### PR TITLE
[fix] flaky SDP tests with Gloo, checking all handles

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Seems to fix https://github.com/facebookresearch/fairscale/issues/498 (see [CI](https://app.circleci.com/pipelines/github/facebookresearch/fairscale?branch=gloo_check_all_handles), I kicked the jobs a couple of times in a row, all green). This is a change which is only needed for Gloo actually, but testing with several backends has other advantages so my take is that it's worth the very small perf hit

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -505,16 +504,15 @@ def _sync_params_and_buffers(self) -> None:
...
-                t, src=self.reference_global_rank, group=self.process_group, async_op=True
+            work_handles.append(
+                dist.broadcast(t, src=self.reference_global_rank, group=self.process_group, async_op=True)
             )
 
-        # Only wait for the last handle, they're inlined in the same CUDA stream
-        if last_work_handle:
-            last_work_handle.wait()
+        # gloo does not guarantee inlining like NCCL, wait for all requests
+        _ = list(filter(lambda x: x.wait(), work_handles))
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:515]: the crux of the diff: Gloo is actually not guaranteeing that the network calls will be sequential (or so I've been told), which means that it's safer to actually check for all handles (not needed for NCCL, but I figured that a backend check here would really be ugly). same below

**Comment by blefaudeux** [comment]
(the broken test is not ShardedDDP, looks like a flaky MoE that @myleott fixed earlier)

**Review by min-xu-ai — APPROVED** [review]

File: fairscale/nn/data_parallel/sharded_ddp.py (hunk 1 above)
Comment (anj-s) [reply fairscale/nn/data_parallel/sharded_ddp.py:515]: is there a perf difference for NCCL? I vaguely remember this change helped with perf but I could have misremembered the PR.

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py:515]: yes there's a small perf hit, you can see all the checks on the trace which take a non-negligible amount of time (like, %), but I think that having something super robust is still worth it for now, until we have a real perf-constrained usecase ? right now the broadcast part is slowing things down so much that this is o()

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py]: (done)

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -540,19 +540,19 @@ def _sync_param_groups(source: List[Dict[Any, Any]], destination: List[Dict[Any,
...
-                    global_src_rank = self.get_global_rank(self.group, src_rank)
-                    last_work_handle = dist.broadcast(
-                        tensor=bucket, src=self._local_to_global_rank[src_rank], group=self.group, async_op=True
+                    work_handles.append(
+                        dist.broadcast(
+                            tensor=bucket, src=self._local_to_global_rank[src_rank], group=self.group, async_op=True
+                        )
                     )
 
         # Only check on the last handle, they're all inlined on the same CUDA stream
```
Comment (anj-s) [inline fairscale/optim/oss.py:554]: Do we need to update this as well?

**Pull request merged**